---
title: "1. Processing Data: 21LT2"
author: "Friederike DÃ¼ndar"
date: "March 2020; updated `r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{1. Processing Data: 21LT2}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval=TRUE, cache=FALSE
)
```

# Background

There are two types of samples from previously *in vitro* sensitized T cells to be compared here:

* T cells presented with the **irrelevant** (WT) antigen, and
* T cells presented with the **tumor antigen**

Both samples contain a mixture of T cells and APC, although the ratio is skewed heavily towards T cells (approximately 10:1).

All samples were processed at the Epigenomics Core (WCM) using the Chromium Single Cell Immune Profiling Solution.
This method allows to obtain both the gene expression profile and T-cell receptor repertoires from the same input sample.
In short, for each input sample, a single batch of barcoded droplets was created followed by cDNA library generation for each cell.
After cDNA amplification, each sample was split into two aliquots, of which one was subjected to T cell enrichment (`Chromium Human T Cell`).
Both aliquots were then sequenced separately, yielding two distinct libraries from the same input samples (identified with separate I7 indices).
Using the `CellRanger` package supplied by 10X Genomics, the Epigenomics Core created two files per sample:

* a count matrix of transcript counts per gene and cell, and
* a table of TRA/TRB clonotypes per cell.

The pipeline includes read alignment to the human genome and removal of barcodes with very low read counts.
The annotation files that were used were `vdj_GRCh38_alts_ensembl-3.1.0-3.1.0` and `GRCh38-3.0.0`. 
The Epigenomics Core ID is: `EC-SC-5166`, which will allow the retrieval of the original Cell Ranger output.

## Data processing

Our data processing encompassed additional quality checks and removal of genes and cells with aberrant transcript numbers and lack of VDJ information.

## Summary

```{r}
library(Klebanoff21LT2)
data(cell_filt)
data(gene_filt)
```


The count matrix provided by Cell Ranger contained  `r print(gene_filt$ori)` genes x `r print(cell_filt$ori)` cells.

1. removed genes without single read in any of the cells
2. removed cells with `subsets_mitochondrial_percent` >= 14%: `r print(cell_filt$mito)`
2. removed cells with `log10_total_features` outside of 3x median: `r print(cell_filt$n_features)`
3. removed cells without clonotype information: `r print(cell_filt$clonotype)`
4. removed cells with 2 or more TRB or TRA clonotypes (= doublets): `r print(cell_filt$clonotype_doubs)`
5. removed genes that were detected in fewer than 5 cells per condition: `r print(gene_filt$clonotype_doubs_min5)` genes

Final DGE matrix with TRA/TRB information for each cell: `r print(gene_filt$clonotype_doubs_min5)` genes x `r print(cell_filt$clonotype_doubs)` cells (across both samples).

# Processing

## Reading in

Reading in the count matrices and VDJ resuls as provided by CellRanger.

```{r libraries, message=FALSE, warning=FALSE}
library(data.table)
library(magrittr)
library(SingleCellExperiment)  
library(ggplot2); theme_set(theme_bw(base_size = 16) )
library(ggrepel)
library(patchwork)
```


```{r readingIn, eval=FALSE}
cell_filt <- list()
gene_filt <- list()

## CLONOTYPE INFO ===========================
samplename <- "21LT2"
data_dir <- "/scratchLocal/frd2007/2018-11_Smita_Tcells/data/21LT2/"
## stored within the tar.gz generated with CellRanger and downloaded from PubShare
clono_info_wt <- reading_in_clonotypes(paste0(data_dir,"WT_VDJ/")) # 21LT2_H-VDJ.tar.gz
clono_info_mut <- reading_in_clonotypes(paste0(data_dir,"MUT_VDJ/")) # 21LT2_R-VDJ.tar.gz

## READ COUNTS ==============================
# list.files(dir.name) # "barcodes.tsv" "genes.tsv"    "matrix.mtx"
wt <- paste0(data_dir, "WT_GEX/outs/filtered_feature_bc_matrix")
mut <- paste0(data_dir, "MUT_GEX/outs/filtered_feature_bc_matrix")

sce.wt <- DropletUtils::read10xCounts(wt, 
  sample.names = paste("WT", samplename, sep = "."),
  version = "auto")
sce.wt$cell <- paste(sce.wt$Sample, 1:ncol(sce.wt), sep = "_")
colnames(sce.wt) <- sce.wt$cell

sce.mut <- DropletUtils::read10xCounts(mut, 
  sample.names = paste("MUT", samplename, sep = "."),
  version = "auto")
sce.mut$cell <-  paste(sce.mut$Sample, 1:ncol(sce.mut), sep = "_")
colnames(sce.mut) <- sce.mut$cell

## prepare cell info 
ci.wt <- prep_cell_info(sce.wt, clono_info_wt)
ci.mut <- prep_cell_info(sce.mut, clono_info_mut)

## COMBINE ------------------------------------
cts <- cbind(counts(sce.mut), counts(sce.wt)) ## keep order!
cell_filt$ori <- ncol(cts)
gene_filt$ori <- nrow(cts)

## combine colData
cd <- rbind(ci.mut, ci.wt)
cd <- cd[colnames(cts),]

## rowData prep
gene_info <- rowData(sce.mut)[, c(1:2)]
names(gene_info) <- c("ensembl_id","gene_symbol")

## make SCE object --------------------------------------------
sce <- SingleCellExperiment(assays = list(counts = cts),
  colData = cd,
  rowData = gene_info[rownames(cts),])

## remove genes with only zero counts----------------------------
keep <- rowSums(counts(sce)) > 0 
sce <- sce[keep,]
dim(sce)
#[1]] 21453  7533
# [1] 21992 11173 re-process, 2020
gene_filt$first_all_zeros <- nrow(sce)

## ADD CHROMOSOME INFO & CELL QC METRICS -------------------
sce <- get_mito(sce)
is.mito <- rowData(sce)$chr == "chrMT"
sce <- add_cell_qc(sce, is.mito)
##! saveRDS(sce, file = "sce_21LT2_unfiltered.rds")
```


## Remove low-quality cells

```{r cellEval, eval=FALSE}
png("qc_per_cell.png", height = 600, width = 700)
p1 <- scater::plotColData(sce, 
  y = "Sample", x = "log10_total_features",
  colour_by = "Sample") +
  theme(legend.position = "bottom") +
  scale_fill_manual(values = c("dodgerblue1","limegreen"))

p2 <- scater::plotColData(sce, 
  y = "Sample", x = "subsets_mitochondrial_percent",
  colour_by = "Sample") +
  theme(legend.position = "none")+
  scale_fill_manual(values = c("dodgerblue1","limegreen"))

p3 <- scater::plotColData(sce,
  y = "log10_total_features", x = "subsets_mitochondrial_percent",
  colour_by = "Sample",  other_fields = "Sample") +
  facet_grid(~Sample) +
  theme(legend.position = "none")+
  scale_fill_manual(values = c("dodgerblue1","limegreen"))

(p1 | p2 ) / p3
dev.off()
```

```{r cellFiltering, eval=FALSE}
options(stringsAsFactors = FALSE)
sce.filt <- sce
rm(sce); gc()
## Mito content ---------------------------------------------------------
high.mito <- scater::isOutlier(sce.filt$subsets_mitochondrial_percent, nmads = 5, type = "higher")
sce.filt[, !high.mito]$subsets_mitochondrial_percent %>% max
# [1] 14.68138
# [1] 19.03831 ## re-process, 2020

### manually defined mitochondrial threshold: 14%
high.mito <- sce.filt$subsets_mitochondrial_percent >= 14
sce.filt <- sce.filt[, !high.mito]
cell_filt$mito <- ncol(sce.filt)

## Number of genes per cell--------------------------------------------
outlier.genes <- scater::isOutlier(sce.filt$detected, nmads = 3, type = "both")
sce.filt[, !outlier.genes]$log10_total_features  %>% summary
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 3.011   3.420   3.498   3.526   3.630   3.809 
## re-process, 2020:
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#   2.220   3.357   3.449   3.446   3.587   3.794 

sce.filt <- sce.filt[, !outlier.genes]
# [1] 21992  8984 ## re-process, 2020
cell_filt$n_features <- ncol(sce.filt)

## immediately remove no-coverage-genes --------------------------------------
keep <- rowSums(counts(sce.filt)) > 0
sce.filt <- sce.filt[keep,]

dim(sce.filt)
#[1]  21430  8984 # re-process, 2020
gene_filt$low_qual_cells <- nrow(sce.filt)

## CELL CYCLE STATUS ---------------------------------------------------------
cc.noNormAllCells <- get_cc_info(sce.filt, print_results = TRUE)
  
# G1  G2M    S 
#3258  223  674 
#################
#      mut   wt
# G1  2433 3582
# G2M  156  290
# S    404  460
###############
#     MUT   WT
#G1  1256 2002
#G2M   75  148
#S    302  372
colData(sce.filt)$cc_phase <- cc.noNormAllCells$phases
colData(sce.filt)$G1score <- cc.noNormAllCells$score$G1
colData(sce.filt)$G2Mscore <- cc.noNormAllCells$score$G2M

## make more sensible rownames ---------------------------
rownames(sce.filt) <- scater::uniquifyFeatureNames(
  rowData(sce.filt)$ensembl_id,
  rowData(sce.filt)$gene_symbol)

!##saveRDS(sce.filt, file = "sce_21LT2_filtered.rds")
```

Counts & genes per cell before filtering:

![](/Users/frd2007/Documents/Projects/2018-10_Smita/chris-klebanoff/Klebanoff21LT2/inst/extdata/figures/qc_per_cell.png)


## QC & Filtering of Genes


```{r geneFiltering, eval=FALSE}
## calculate qc metrics per gene across all the cells of the same sample
gnqc.list <- do_gene_qc(sce.filt, return_df = TRUE)
gn.qc <- gnqc.list$gene_qc

png("qc_genes.png", width=1000, height = 750)
p1 <- ggplot(data = data.frame(gn.qc), aes(x = subsets_WT.21LT2_mean,
  y = subsets_WT.21LT2_detected)) + 
  geom_point(shape=1, size=.5) +
  ggtitle("WT 21LT2") +
  scale_x_log10() + scale_y_log10() +
  ylab("% cells with non-zero expression") + xlab("mean counts")
p1 <- p1 + 
  geom_point(data = subset(data.frame(gn.qc), Gene %in% gnqc.list$hk_genes), 
    color = "red",shape = 22, size = 4) 

p2 <- ggplot(data = data.frame(gn.qc), aes(x = subsets_MUT.21LT2_mean,
  y = subsets_MUT.21LT2_detected)) + geom_point(shape=1, size=.5) + 
  ggtitle("MUT 21LT2") +
  scale_x_log10() + scale_y_log10() +
  ylab("% cells with non-zero expression") + xlab("mean counts")
p2 <- p2 + 
  geom_point(data = subset(data.frame(gn.qc), Gene %in% gnqc.list$hk_genes), 
    color = "red",shape = 22, size = 4) 

p3 <- ggplot(data = data.frame(gn.qc),
  aes(x = subsets_MUT.21LT2_detected, y = subsets_WT.21LT2_detected)) +
  geom_point(shape=1, size=.5, alpha =.5) +
  scale_x_log10() + scale_y_log10()
p3 <- p3 +
  geom_point(data = subset(data.frame(gn.qc), Gene %in% gnqc.list$hk_genes), 
    color = "red",shape = 22, size = 4) 
(p1 | p2) / p3
dev.off()

png("qc_highestExpressedGenes.png", width= 550, height = 500)
scater::plotHighestExprs(sce.filt) 
dev.off()
```

Dropout rate assessment:

![](/Users/frd2007/Documents/Projects/2018-10_Smita/chris-klebanoff/Klebanoff21LT2/inst/extdata/figures/qc_genes.png)

Highest expressed genes:

![](/Users/frd2007/Documents/Projects/2018-10_Smita/chris-klebanoff/Klebanoff21LT2/inst/extdata/figures/qc_highestExpressedGenes.png)

```{r rm_low_qual_genes, eval=FALSE}
keep_genes <-  do_gene_qc(sce.filt, min.cells = 5, return_df = FALSE)
sce.filt <- sce.filt[gnqc.list$keep_genes,]

dim(sce.filt)
# > dim(sce.filt)
# [1] 18776  8984 # re-process, 2020
gene_filt$low_qual_genes <- nrow(sce.filt)

!##saveRDS(sce.filt, file = "sce_21LT2_filtered.rds")
```


## Normalization and Dimensionality Reductions

Following filtering, we're going to normalize, adjusting for batch effect (=Sample) and performing dimensionality reduction.

```{r normalization, eval=FALSE}
library(scran)
library(batchelor)
library(scater)

## size factor --------------------------------------
#message("calculating the size factors")
qckclst <- quickCluster(sce.filt, method = "igraph",
  min.mean = 0.1)
sce.filt <- computeSumFactors(sce.filt, min.mean=0.1, cluster = qckclst)
sce.filt <- scater::logNormCounts(sce.filt)
##!saveRDS(sce.filt, file = "sce_21LT2_filtered.rds")

## variance modelling -------------------------------
dec.filt <- modelGeneVar(sce.filt, block=sce.filt$Sample)
chosen.hvgs <- dec.filt$bio > 0
table(chosen.hvgs)
#chosen.hvgs
#FALSE  TRUE
#10025  8751

#--- merging ---#
set.seed(01001001)
sce.merged <- correctExperiments(sce.filt, 
    batch=sce.filt$Sample, 
    subset.row=chosen.hvgs)

sce.merged
#class: SingleCellExperiment 
#dim: 8565 9015 
#metadata(1): merge.info
#assays(3): reconstructed counts logcounts
#rownames(8565): MIR1302-2HG AL627309.1 ... AC004556.1 AC240274.1
#rowData names(4): ensembl_id gene_symbol chr rotation
#colnames(9015): MUT.21LT22 MUT.21LT23 ... WT.21LT26363 WT.21LT26364
#colData names(18): batch Barcode ... G1score G2Mscore
#reducedDimNames(1): corrected
#spikeNames(0):
#altExpNames(0):
##! saveRDS(sce.merged, file = "sce_batchCorrected.rds")
```


## Remove cells without clonotype information

```{r clonotype_filtering, eval=FALSE}
## REMOVING CELLS WITHOUT CLONOTYPE----------------------------
sce.filt <- sce.filt[,which(!is.na(sce.filt$cdr3s_aa))]
dim(sce.filt)
# [1] 18776  5172
cell_filt$clonotype <- ncol(sce.filt)

## frequencies of chains, i.e. counting number of TRA/TRB entries within cdr3s_aa column
sce.filt <- add_chain_counts(sce.filt)
## frequency of clonotypes, i.e. counting individual cdr3s_aa entries
sce.filt$freq_per_Sample <- add_frequencies(sce.filt, count_what = "cdr3s_aa", mode = "per.sample")
sce.filt$freq_across_all <- add_frequencies(sce.filt, count_what = "cdr3s_aa", mode = "all")

## REMOVING PUTATIVE DOUBLETS --------------------------------
png("relationship_TRB_features.png", width = 850)
cd <- colData(sce.filt)
p1 <- ggplot(as.data.frame(cd), aes(x = numTRA, 
  y = log10_total_features, color = subsets_mitochondrial_percent)) + 
  ggbeeswarm::geom_quasirandom(size = 1, alpha = .5, shape = 1) +
  theme(legend.position = "bottom") +
  scale_colour_viridis_c()

p2 <- ggplot(as.data.frame(cd), aes(x = numTRB, 
  y = log10_total_features, color = subsets_mitochondrial_percent)) + 
  ggbeeswarm::geom_quasirandom(size = 1, alpha = .5, shape = 1) +
 # ggtitle(" 'Cells' with more than 1 TRB tend to have increased numbers of counts") +
  theme(legend.position = "bottom") +
  scale_colour_viridis_c()

pw <- p1 + p2
pw + plot_annotation(
  title = 'Are multipe TRA/TRB per cell indicative of technical doublets?'
  #subtitle = 'These 3 plots will reveal yet-untold secrets about our beloved data-set',
  #caption = 'Disclaimer: None of these plots are insightful'
)
dev.off()

## after looking at these plots, I'm fairly certain that removing cells with more 
## than 2 TRB is benefitial to avoid doublets
sce.filt <- sce.filt[, sce.filt$numTRB < 2 & sce.filt$numTRA < 2]
## remove all-newly-only-zero-genes
keep <- rowSums(as.matrix(counts(sce.filt))) > 0
sce.filt <- sce.filt[keep, ]
cell_filt$clonotype_doubs <- ncol(sce.filt)
gene_filt$clonotype_doubs <- nrow(sce.filt)

dim(sce.filt)
#[1] [1] 18382  4392

keep <- do_gene_qc(sce.filt, min.cells = 5, return_df = FALSE)
gene_filt$clonotype_doubs_min5 <- nrow(sce.filt)
```

![](/Users/frd2007/Documents/Projects/2018-10_Smita/chris-klebanoff/Klebanoff21LT2/inst/extdata/figures/relationship_TRB_features.png)

## Clustering

```{r clustering, eval=FALSE}
kp <- rownames(sce.merged) %in% rownames(sce.filt)
sce.merged <- sce.merged[kp, colnames(sce.filt)]
##  clustering using the batch-corrected values ---------------
g <- buildSNNGraph(sce.merged, use.dimred="corrected")
clusters <- igraph::cluster_louvain(g)
sce.merged$cluster <- factor(clusters$membership)

#--- dimensionality-reduction ---#
sce.merged <- runTSNE(sce.merged, dimred="corrected", external_neighbors=TRUE)
sce.merged <- runUMAP(sce.merged, dimred="corrected", external_neighbors=TRUE)

png("umap_samples.png")
p1 <- plotReducedDim(sce.merged, dimred="UMAP", colour_by="Sample") +
    scale_fill_manual(values = c("dodgerblue1","limegreen"))
p2 <- plotReducedDim(sce.merged, dimred="UMAP", colour_by="cluster",
   other_fields = "Sample") +
  facet_grid(~Sample)
p1 / p2
dev.off()

##!saveRDS(sce.merged, file = "sce_batchCorrected.rds")

reducedDims(sce.filt) <- reducedDims(sce.merged)
sce.filt$cluster <- sce.merged$cluster

##! saveRDS(sce.filt, file = "sce_21LT2_clonotypefiltered.rds")
##! saveRDS(sce.merged, file = "sce_21LT2_batchCorrected_clonotypefiltered.rds")
```

# Results

```{r}
sce.filt <- load_21LT2filt()
```

```{r umap_overview, fig.show='hold', fig.width = 20, fig.height = 18, echo=FALSE}
goi <- c("Sample","subsets_mitochondrial_percent", "log10_total_features", "cluster",
         "numTRA", "numTRB", "freq_per_Sample", "freq_across_all", "cc_phase")

pl <- lapply(goi, function(x){
  scABC2::plot_reducedDim_from_sce(sce.filt,
    which_reddim="UMAP",
    color_by=x, exprs_values="logcounts",
    size=.8, alpha = .7) + 
  ggtitle(x) + xlab("UMAP 1") + ylab("UMAP 2") +
    theme(legend.position = "bottom")
})

ABCutilities::MultiPlotList(pl, cols = 3)
```

## Monocytes

```{r umap_monocytes, fig.show='hold', echo=FALSE, fig.width = 12, fig.height = 5}
goi <- c("CD83","CD86", "CD40")

pl <- lapply(goi, function(x){
  P <- scABC2:::plot_reducedDim_from_sce(sce.filt,
    which_reddim="UMAP",
    color_by=x, exprs_values="logcounts",
    size=.7, alpha = .7, remove_rug=TRUE,
    label_by = "cluster", label_size = 6)
  P <- P +  ggtitle(paste("Monocytes:", x)) + xlab("UMAP 1") + ylab("UMAP 2") +
    theme(legend.position = "bottom")
  return(P)
#  print(P)
})

ABCutilities::MultiPlotList(pl, cols = 3)
```

## NK cells, DCs, B cells 

```{r umap_otherImmuneCells,echo=FALSE, fig.width = 10, fig.height = 10}
scABC2:::plot_reducedDim_from_sce(sce.filt, which_reddim="UMAP",
  color_by="GZMB", exprs_values="logcounts",
  size=1, alpha = .7) + 
  ggtitle("CTL / NK cells: GZMB") + xlab("UMAP 1") + ylab("UMAP 2") +
  theme(legend.position = "bottom")

scABC2:::plot_reducedDim_from_sce(sce.filt,
  which_reddim="UMAP", 
  color_by="CD80", exprs_values="logcounts",
  size=.7, alpha = .7, label_by = "cluster", label_size = 6) + 
  ggtitle("mature DCs: CD80") + xlab("UMAP 1") + ylab("UMAP 2") +
  theme(legend.position = "bottom")

scABC2:::plot_reducedDim_from_sce(sce.filt,
  which_reddim="UMAP", color_by="CD86",
  exprs_values="logcounts", size=.7, label_by = "cluster", label_size = 6) +
  ggtitle("mature DCs: CD86")+ xlab("UMAP 1") + ylab("UMAP 2") +
  theme(legend.position = "bottom")

scABC2:::plot_reducedDim_from_sce(sce.filt, which_reddim="UMAP",
  color_by="CD19", exprs_values="logcounts", 
  size=.7, alpha = .7,label_by = "cluster", label_size = 6) +
  ggtitle("B cells: CD19")+ xlab("UMAP 1") + ylab("UMAP 2") +
  theme(legend.position = "bottom")
```

## T cells

```{r umap_tcells, echo=FALSE, fig.width = 10, fig.height = 10}
pl <- lapply(c("CD4","CD3E","CD8A","CD8B"), function(x){
  P <- scABC2:::plot_reducedDim_from_sce(sce.filt,
    which_reddim="UMAP", color_by=x, 
    exprs_values="logcounts", 
    size=.7, alpha= .7,
    label_by = "cluster", label_size = 6) 
  P <- P + ggtitle(paste("T cells:", x)) + xlab("UMAP 1") + ylab("UMAP 2") +
    theme(legend.position = "bottom")
  return(P)
})

ABCutilities::MultiPlotList(pl, cols = 2)
```

## Markers of activation

```{r umap_activation_markers, fig.width = 15, fig.height = 10, echo=FALSE}
goi <- c("Sample","IFNG","TNF", "IL2","TNFRSF9", "TNFRSF4")

pl <- lapply(goi, function(x){
  P <- scABC2:::plot_reducedDim_from_sce(sce.filt,
    which_reddim="UMAP",
    color_by=x, exprs_values="logcounts",
    size=.7, alpha = .7,label_by = "cluster", label_size = 6)
  P <- P +  ggtitle( x) + xlab("UMAP 1") + ylab("UMAP 2") +
    theme(legend.position = "bottom")
  return(P)
})

ABCutilities::MultiPlotList(pl, cols = 3)
```

| Cluster | Identity |
|----------|---------|
| 1 | CD4 |
| 2 | Monocytes |
| 3 | **activated CD8** |
| 4 | CD4 |
| 5 | unclear, tiny |
| 6 | CD8 (in G2M) |
| 7 | unclear, tiny |
| 8 | CD8 |
| 9 | CD8|
|10| CD8 |


# Identify shared clonotypes

= clonotypes that are present in both conditions, i.e. MUT and WT.

Get a data.table with unique clonotypes per sample:

```{r}
## clonotype frequencies
clono_freq <- colData(sce.filt)[, c("Sample" ,"cdr3s_aa")] %>% 
  as.data.frame %>% data.table(., keep.rownames = TRUE) %>% 
  .[!is.na(cdr3s_aa), .N, c("cdr3s_aa","Sample")] 
setorder(clono_freq, N)
```

```
> clono_freq
                                       cdr3s_aa    Sample   N
   1:    TRA:CAAPLGGNTGKLIF;TRB:CASSRYSGGTYGYTF MUT.21LT2   1
   2:    TRA:CLVGDSNSGNTPLVF;TRB:CASSYSGRNYGYTF MUT.21LT2   1
   3:                  TRB:CASSLKGLMNLRREVGEQYF MUT.21LT2   1
   4:         TRA:CAVNPGTASKLTF;TRB:CASSLQLEQYF MUT.21LT2   1
   5:      TRA:CAASSYSGGGADGLTF;TRB:CASKRADTQYF MUT.21LT2   1
  ---                                                        
3295:  TRA:CAGINAGGTSYGKLTF;TRB:CASSLVPGDAYEQYF MUT.21LT2  99
3296:                       TRB:CASSLVPGDAYEQYF  WT.21LT2 101
3297:  TRA:CAGINAGGTSYGKLTF;TRB:CASSLVPGDAYEQYF  WT.21LT2 116
3298: TRA:CAGNTGTASKLTF;TRB:CASSGLAGGPVSGANVLTF MUT.21LT2 124
3299: TRA:CAGNTGTASKLTF;TRB:CASSGLAGGPVSGANVLTF  WT.21LT2 154
```


```{r clonotype_frequencies, fig.width = 7}
ggplot(clono_freq, aes(x = Sample, y = N, key = cdr3s_aa, fill = Sample)) +
  ggbeeswarm::geom_quasirandom(size = 3, shape = 21) +
  ggrepel::geom_label_repel(data = clono_freq[N>30], aes(label = cdr3s_aa), box.padding = 1) +
  scale_fill_manual(values = c("lightskyblue1","limegreen")) +
  coord_flip() + ylab("# cells") + xlab("") +
  ggtitle("Clonotype frequencies") +
  theme(legend.position = "bottom")
```


Identify clonotypes that are present in both sample types:

```{r}
ct <- dcast(clono_freq, cdr3s_aa ~ Sample, value.var = "N") %>% .[!is.na(MUT.21LT2) & !is.na(WT.21LT2)]
```

There are `r nrow(ct)` clonotypes that are found in both MUT and WT samples.

```{r}
## formatting the TRA/TRB notations
## will only work if there's just one TRA
ct[, TRA := gsub(";*TRB:[A-Z]+", "", cdr3s_aa)]
ct[, TRA := ifelse(TRA == "", NA, TRA)]
ct[, TRB := gsub(".*(TRB:[A-Z]+)", "\\1", cdr3s_aa)]
ct[, TRB := ifelse(grepl("^TRA", TRB), NA, TRB)] # if only TRB was present, I need to fill in the NA

setorder(ct, -MUT.21LT2, -WT.21LT2 )
```

```
> ct[]
                                      cdr3s_aa MUT.21LT2 WT.21LT2                     TRA                     TRB
 1:  TRA:CAGNTGTASKLTF;TRB:CASSGLAGGPVSGANVLTF       124      154       TRA:CAGNTGTASKLTF TRB:CASSGLAGGPVSGANVLTF
 2:   TRA:CAGINAGGTSYGKLTF;TRB:CASSLVPGDAYEQYF        99      116    TRA:CAGINAGGTSYGKLTF     TRB:CASSLVPGDAYEQYF
 3:                        TRB:CASSLVPGDAYEQYF        61      101                    <NA>     TRB:CASSLVPGDAYEQYF
 4:        TRA:CIVRGGKLIF;TRB:CASSLALGYLNTEAFF        25       35          TRA:CIVRGGKLIF    TRB:CASSLALGYLNTEAFF
 5:         TRA:CAMRANDYKLSF;TRB:CASSLRADSEQYF        17       14        TRA:CAMRANDYKLSF       TRB:CASSLRADSEQYF
 6:       TRA:CAVRPRNDYKLSF;TRB:CASAVWGVADIQYF        15       18       TRA:CAVRPRNDYKLSF      TRB:CASAVWGVADIQYF
 7:        TRA:CGAVGNTGKLIF;TRB:CAISGGRSSYEQYF        14       18        TRA:CGAVGNTGKLIF      TRB:CAISGGRSSYEQYF
 8:                    TRB:CASSGLAGGPVSGANVLTF        12       14                    <NA> TRB:CASSGLAGGPVSGANVLTF
```


## Dimensionality reduction on the shared clonotypes only

UMAP of **shared clonotypes**

```{r adjusting_colData}
sce.shared <- sce.filt[, sce.filt$cdr3s_aa %in% ct$cdr3s_aa]
#dim(sce.shared)
#17107  1139
```

There are `r ncol(sce.shared)` cells in the SCE object with clonotypes that are present in both samples.


```{r shared_UMAP1}
set.seed(123)
sce.shared <- scater::runUMAP(sce.shared, 
  exprs_values = "logcounts",
  dimred = "corrected")
```

```{r umap_sharedClonotypes,  fig.show='hold', fig.height = 4.5, fig.width = 5, echo=FALSE}
scABC2::plot_reducedDim_from_sce(sce.shared,
  which_reddim = "UMAP", color_by = "Sample", size = 1) +
  ggtitle("UMAP of cells with shared clonotypes") +
  theme(legend.position = "bottom")
```

When coloring the cells by the total number of genes detected, it becomes apparent that the distinct cluster on the left is probably composed of droplets where more than one cell was captured (= doublets).

```{r umap_sharedClonotypes2,  fig.show='hold', fig.height = 4.5, fig.width = 5, echo=FALSE}
## total counts --> doublet cluster becomes obvious
scABC2::plot_reducedDim_from_sce(sce.shared,
  which_reddim = "UMAP",
  color_by = "log10_total_features", set_color=FALSE) +
  scale_color_viridis_c() +
  theme(legend.position = "bottom")
```

When assessing markers of T cells (CD3) and monocytes (CD84), it seems as if these are doublets generated by the capture of two different cell types ince CD3 is lower than in most other cells and CD84 is higher.
In addition, these are definitely not the activated T cells as indicated by IFNg and IL2.

```{r umap_monocyte_diagnostics, fig.show='hold', fig.height = 6, fig.width = 7, echo=FALSE}
library(patchwork)
p1 <- scABC2::plot_reducedDim_from_sce(sce.shared,
  which_reddim = "UMAP", color_by = "IFNG", exprs_values = "logcounts", size = .7, alpha = .7, set_color=FALSE) +
  scale_color_viridis_c() + theme(legend.position = "top")
p2 <- scABC2::plot_reducedDim_from_sce(sce.shared, 
  which_reddim = "UMAP", color_by = "IL2", exprs_values = "logcounts", size = .7, alpha = .7,set_color=FALSE) +
  scale_color_viridis_c() + theme(legend.position = "top")
p3 <- scABC2::plot_reducedDim_from_sce(sce.shared, 
  which_reddim = "UMAP", color_by = "CD86", exprs_values = "logcounts", size = .7, alpha = .7 ,set_color=FALSE) +
  scale_color_viridis_c() + theme(legend.position = "top")
p4 <- scABC2::plot_reducedDim_from_sce(sce.shared, 
  which_reddim = "UMAP", color_by = "CD3E", exprs_values = "logcounts", size = 1,set_color=FALSE) +
  scale_color_viridis_c() + theme(legend.position = "top")

(p3 + p4) / (p1 + p2)
```

## Removal of suspected doublets

Let's remove them using the UMAP coordinates.

```{r removing_doublets}
umap.dt <- reducedDim(sce.shared, "UMAP") %>% data.table(., keep.rownames = TRUE)
outs <- umap.dt[V1 < -10]$rn
sce.shared$umap_group <- ifelse(rownames(colData(sce.shared)) %in% outs, "doublets", "tcells")
table(sce.shared$umap_group)
```

```{r fig.show='hold', fig.height = 4.5, fig.width = 5, echo=FALSE}
scABC2::plot_reducedDim_from_sce(sce.shared, 
  which_reddim = "UMAP", color_by = "umap_group") +
  theme(legend.position = "top")
```

```{r shared_UMAP2, eval=FALSE}
## remove doublets from the SCE
sce.shared <- sce.shared[, sce.shared$umap_group != "doublets"]
dim(sce.shared)
#[1] 17107  1084
set.seed(123)
sce.shared <- scater::runUMAP(sce.shared, exprs_values = "logcounts",
  dimred = "corrected")
colnames(reducedDim(sce.shared, "UMAP")) <- c("UMAP_1","UMAP_2")

##!saveRDS(sce.shared, file = "~/Documents/Projects/2018-10_Smita/data/sce_sharedClonotypes_Mar2020.RDS")
```

This is what the newly re-run UMAP looks like after removing the suspected doublets:

```{r echo=FALSE}
sce.shared <- load_21LT2shared()
```

```{r umap_after_doublet_removal, fig.show='hold', fig.height = 6.5, fig.width = 7.5, echo=FALSE}
library(patchwork)
p1 <- scABC2::plot_reducedDim_from_sce(sce.shared,
  which_reddim = "UMAP", color_by = "IFNG", exprs_values = "logcounts", size = .7, alpha = .7, set_color=FALSE) +
  scale_color_viridis_c() + theme(legend.position = "top")
p2 <- scABC2::plot_reducedDim_from_sce(sce.shared, 
  which_reddim = "UMAP", color_by = "IL2", exprs_values = "logcounts", size = .7, alpha = .7,set_color=FALSE) +
  scale_color_viridis_c() + theme(legend.position = "top")
p3 <- scABC2::plot_reducedDim_from_sce(sce.shared, 
  which_reddim = "UMAP", color_by = "CD86", exprs_values = "logcounts", size = .7, alpha = .7 ,set_color=FALSE) +
  scale_color_viridis_c() + theme(legend.position = "top")
p4 <- scABC2::plot_reducedDim_from_sce(sce.shared, 
  which_reddim = "UMAP", color_by = "CD3E", exprs_values = "logcounts", size = 1,set_color=FALSE) +
  scale_color_viridis_c() + theme(legend.position = "top")

(p3 + p4) / (p1 + p2)
```

Need to clean up; i.e. some cells that were removed will have led to the wiping out of individual samples, i.e. the pool of shared clonotypes has been altered.

## Re-calculate the clonotype frequencies after the removal of the doublets

```{r shared_ct_after_doublet_removal, eval=FALSE}
shared_clonotypes <- colData(sce.shared)[, c("Sample" ,"cdr3s_aa")] %>% 
    as.data.frame %>% data.table(., keep.rownames = TRUE) %>% 
    .[!is.na(cdr3s_aa), .N, c("cdr3s_aa","Sample")] %>%
  dcast(., cdr3s_aa ~ Sample, value.var = "N") %>% .[!is.na(MUT.21LT2) & !is.na(WT.21LT2)]
shared_clonotypes[, TRA := gsub(";*TRB:[A-Z]+", "", cdr3s_aa)]
shared_clonotypes[, TRA := ifelse(TRA == "", NA, TRA)]
shared_clonotypes[, TRB := gsub(".*(TRB:[A-Z]+)", "\\1", cdr3s_aa)]
shared_clonotypes[, TRB := ifelse(grepl("^TRA", TRB), NA, TRB)] # if only TRB was present, I need to fill in the NA
setorder(shared_clonotypes, -MUT.21LT2, -WT.21LT2 )

## rm those cells that are no longer shared across conditions
sce.shared <- sce.shared[, sce.shared$cdr3s_aa %in% unique(shared_clonotypes$cdr3s_aa)]
dim(sce.shared)
#17107  1080
```

## Adding the clonotype IDs

```{r clonotype_consolidation, eval=FALSE}
## extract all shared clonotypes
tt.ori <- data.table(cdr3s_aa = unique(shared_clonotypes$cdr3s_aa))

## work on a backup
tt.ori2 <- copy(tt.ori)

## split the individual entries per clonotype -- I tested in a dry run how many
## columns I had to assign for this; 2 is sufficient for this case -- might not always
## be the case!!
tt.ori2[, c("T1","T2"):=tstrsplit(cdr3s_aa, ";")]

## count TRA & TRB
tt.ori2[, num_TRA := stringr::str_count(cdr3s_aa, "TRA")]
tt.ori2[, num_TRB := stringr::str_count(cdr3s_aa, "TRB")]

## extract those where we have one or two TRA, but just one TRB
## (we will add the single TRA| single TRB instances later)
tt.good <- tt.ori2[ (num_TRA == 1|num_TRA ==2) & num_TRB == 1,
                    c("cdr3s_aa","T1","T2", "num_TRA","num_TRB"),
                    with=FALSE] %>% unique
setorder(tt.good, T1, T2)
## assign arbitrary ID
maxn <- nrow(tt.good)
tt.good[, id := paste0("C",1:maxn)]

## obtain a skinny list of IDs and original cdr3s_aa entries
## coupled to the individual TRA and TRB
tt.good <- melt(tt.good[, -c("num_TRA","num_TRB"),with=FALSE], 
                id.vars = c("id","cdr3s_aa")) %>% 
  .[!is.na(value)] %>% .[, -"variable", with = FALSE] %>% unique
#      id                      cdr3s_aa                   value
#  1:  C1   TRA:CAAASGGYQKVTF;TRB:CAWRTSGTYEQYF       TRA:CAAASGGYQKVTF
#  2:  C2   TRA:CAAFSGSARQLTF;TRB:CASSLNSGGYGYTF       TRA:CAAFSGSARQLTF
#  3:  C3   TRA:CAAPDPHSGNTPLVF;TRB:CASTIVGYEQYF     TRA:CAAPDPHSGNTPLVF
#  4:  C4   TRA:CAASADDYKLSF;TRB:CASSQVQSGSGNTIYF        TRA:CAASADDYKLSF
#  5:  C5   TRA:CAASAHRYSGGSNYKLTF;TRB:CASSSGTGNSPLHF  TRA:CAASAHRYSGGSNYKLTF

## add the cases with only one TRA or only one TRB (would have messed up the
## workflow above otherwise)
tt.quest <- tt.ori2[num_TRA < 1 | num_TRB < 1,
  c("cdr3s_aa","T1","T2", "num_TRA","num_TRB"),with=FALSE] %>% unique

## combine good and questionable CDRs
tt.combi <- merge(tt.good, tt.quest, by.x = "value", by.y = "T1", all = TRUE)
setorder(tt.combi, value)

## the instances from tt.quest without a match in tt.good will have an NA in the ID, 
## which will have to be rectified
n_na <- nrow(tt.combi[is.na(id)])
tt.combi[is.na(id), id := paste0("C", (maxn+1):(maxn + n_na))]

## keep the original entries from either tt.good or tt.quest; the result will be
## the longest possible combination, i.e. TRA:CAGINAGGTSYGKLTF will belong to TRA:CAGINAGGTSYGKLTF;TRB:CASSLVPGDAYEQYF
tt.combi <- tt.combi[, cdr3s_consolidated := ifelse(is.na(cdr3s_aa.x), cdr3s_aa.y, cdr3s_aa.x)]

## clean up
tt.combi <- unique(tt.combi[, c("id","cdr3s_aa.x","cdr3s_aa.y", "cdr3s_consolidated"), with=FALSE])
setnames(tt.combi, c("cdr3s_aa.x","cdr3s_aa.y"), c("cdr3s_aa.good", "cdr3s_aa.single"))

## bring back the original CDRs
tt.combi <- merge(tt.ori, tt.combi, by.x = "cdr3s_aa" ,by.y = "cdr3s_aa.good", all.x = TRUE) %>% 
  merge(., tt.combi, by.x = "cdr3s_aa", by.y = "cdr3s_aa.single", all.x = TRUE)
tt.combi[, id := ifelse(is.na(id.x), id.y, id.x)]
tt.combi[, cdr3s_consolidated := ifelse(is.na(cdr3s_consolidated.x), cdr3s_consolidated.y, cdr3s_consolidated.x)]
tt.combi <- unique(tt.combi[, c("cdr3s_aa", "id","cdr3s_consolidated"), with=FALSE])

clonotype_ids <- copy(tt.combi)
clonotype_ids$Patient <- "21LT2"
clonotype_ids <- shared_clonotypes[clonotype_ids, on = "cdr3s_aa"] 
```

```
> head(clonotype_ids)
                                    cdr3s_aa MUT.21LT2 WT.21LT2                    TRA                  TRB id
1:       TRA:CAAASGGYQKVTF;TRB:CAWRTSGTYEQYF         1        1      TRA:CAAASGGYQKVTF    TRB:CAWRTSGTYEQYF C1
2:      TRA:CAAFSGSARQLTF;TRB:CASSLNSGGYGYTF         3        3      TRA:CAAFSGSARQLTF   TRB:CASSLNSGGYGYTF C2
3:      TRA:CAAPDPHSGNTPLVF;TRB:CASTIVGYEQYF         1        1    TRA:CAAPDPHSGNTPLVF     TRB:CASTIVGYEQYF C3
4:     TRA:CAASADDYKLSF;TRB:CASSQVQSGSGNTIYF         1        1       TRA:CAASADDYKLSF TRB:CASSQVQSGSGNTIYF C4
5: TRA:CAASAHRYSGGSNYKLTF;TRB:CASSSGTGNSPLHF         1        1 TRA:CAASAHRYSGGSNYKLTF   TRB:CASSSGTGNSPLHF C5
6:     TRA:CAASIGGGAQKLVF;TRB:CASRSDRGENIQYF         3        1     TRA:CAASIGGGAQKLVF   TRB:CASRSDRGENIQYF C6
                          cdr3s_consolidated Patient
1:       TRA:CAAASGGYQKVTF;TRB:CAWRTSGTYEQYF   21LT2
2:      TRA:CAAFSGSARQLTF;TRB:CASSLNSGGYGYTF   21LT2
3:      TRA:CAAPDPHSGNTPLVF;TRB:CASTIVGYEQYF   21LT2
4:     TRA:CAASADDYKLSF;TRB:CASSQVQSGSGNTIYF   21LT2
5: TRA:CAASAHRYSGGSNYKLTF;TRB:CASSSGTGNSPLHF   21LT2
6:     TRA:CAASIGGGAQKLVF;TRB:CASRSDRGENIQYF   21LT2
```

```{r adding_clonotype_ID, eval=FALSE}
## add clonotype ID and consolidated info to the SCE object
#data("clonotype_ids", package = "Klebanoff21LT2") ## loading unique identifiers of the shared clonotypes

## formatting colData
cd.ct <-  data.table(as.data.frame(colData(sce.shared)), keep.rownames = TRUE)
cd.ct <- clonotype_ids[cd.ct, on = "cdr3s_aa"]
cd.ct <-  DataFrame(cd.ct[, -c("rn", "MUT.21LT2","WT.21LT2")], row.names = cd.ct$rn)
colData(sce.shared) <- cd.ct[colnames(sce.shared),]

# add the info to the big data set, too, in case we ever want to go back
cd.all <- data.table(as.data.frame(colData(sce.filt)), keep.rownames = TRUE)
cd.all <- clonotype_ids[cd.all, on = "cdr3s_aa"]
cd.all <-  DataFrame(cd.all[, -c("rn", "MUT.21LT2","WT.21LT2")], row.names = cd.all$rn)
colData(sce.filt) <- cd.all[colnames(sce.filt),]

## determine cells where we have at least 2 instances per condition
testfit <- clonotype_ids[ MUT.21LT2 > 1 & WT.21LT2 > 1]$id %>% unique
## adjust the frequencies of the TRB.TRA
sce.shared$freq_per_Sample <- add_frequencies(sce.shared, count_what = "id", mode = "per.sample")
sce.shared$freq_across_all <- add_frequencies(sce.shared, count_what = "id",mode = "all")
sce.shared$fit_for_test <- ifelse(sce.shared$id %in% testfit, TRUE, FALSE)
```


-----------------------------

# Saving

```{r saving_internally_stored_data, eval=FALSE}
## for data stored in the data/ folder of the package
usethis::use_data(cell_filt)
usethis::use_data(gene_filt)


usethis::use_data(shared_clonotypes, overwrite = TRUE)
usethis::use_data(clonotype_ids, overwrite = TRUE)
```

```{r saving_datasets_for_retrieval, eval=FALSE}
## the indidivudal components of the sce data set will be stored on the Box because
## they are too big to be part of the package
## I therefore save them in a separate folder (sce_components_*) which I will then move to 
## the Box (Folder: Klebanoff_SingleCellData). In the Box, I then navigate to the folder and retrieve the *direct* shared 
## links (--> "Share file" --> "Enable shared link" --> "Settings/Properties" --> get the *direct*
## link (should end with .rds). That link is then stored in a text file, which I keep
## in the inst/extdata folder and will be used by the function to retrieve the data
## (see load_sce() in R/data_sce.R)
saveRDS(colData(sce.filt), file = "~/Documents/Projects/2018-10_Smita/2020-03_reprocessing_21LT2/sce_components_21LT2/colData_sce_filtered.rds")
saveRDS(rowData(sce.filt), file = "sce_components_21LT2/rowData_sce_filtered.rds")
saveRDS(reducedDims(sce.filt), file = "sce_components_21LT2/redDims_sce_filtered.rds")
saveRDS(metadata(sce.filt), file = "sce_components_21LT2/metadata_sce_filtered.rds")
for(i in assayNames(sce.filt)){
  saveRDS(assay(sce.filt, i), file = paste("sce_components_21LT2/assayData", i, "sce_filtered.rds", sep = "_"))
}


## saving the batch-corrected merged SCE
saveRDS(colData(sce.merged), file = "sce_components_21LT2/colData_sceMerged_filtered.rds")
saveRDS(rowData(sce.merged), file = "sce_components_21LT2/rowData_sceMerged_filtered.rds")
saveRDS(reducedDims(sce.merged), file = "sce_components_21LT2/redDims_sceMerged_filtered.rds")
saveRDS(metadata(sce.merged), file = "sce_components_21LT2/metadata_sceMerged_filtered.rds")
for(i in assayNames(sce.merged)){
  saveRDS(assay(sce.merged, i), file = paste("sce_components_21LT2/assayData", i, "sceMerged_filtered.rds", sep = "_"))
}

## saving the colData for shared clonotypes without suspected doublets
saveRDS(colData(sce.shared), file = "~/Documents/Projects/2018-10_Smita/2020-03_reprocessing_21LT2/sce_components_21LT2/colData_sceShared_filtered.rds")
  saveRDS(rowData(sce.shared), file = "~/Documents/Projects/2018-10_Smita/2020-03_reprocessing_21LT2/sce_components_21LT2/rowData_sceShared_filtered.rds")
saveRDS(reducedDims(sce.shared), file = "~/Documents/Projects/2018-10_Smita/2020-03_reprocessing_21LT2/sce_components_21LT2/redDims_sceShared_filtered.rds")
saveRDS(metadata(sce.shared), file = "~/Documents/Projects/2018-10_Smita/2020-03_reprocessing_21LT2/sce_components_21LT2/metadata_sceShared_filtered.rds")
saveRDS(assay(sce.shared, "logcounts"), 
  file = "~/Documents/Projects/2018-10_Smita/2020-03_reprocessing_21LT2/sce_components_21LT2/assayData_logcounts_sceShared_filtered.rds")
```

This is what the storage file should like like:

```{r}
system.file("extdata",  "sce_storage_21LT2.txt", package = "Klebanoff21LT2") %>%
  read.table 
```

There are three storages files:

| File name | Content | Function for loading | 
|-----------|----------|---------------------|
| `sce_storage_21LT2.txt` | data for reconstituting `sce.filt` | `load_21LT2filt()` |
| `sce_storage_21LT2Merged.txt` | data for `sce.merged`, i.e. batch-corrected dim. reds | `load_21LT2merged()` |
| `sce_storage_21LT2Shared.txt` | data for `sce.shared`, i.e. those cells that have clonotypes that are present in both MUT and WT condition | `load_21LT2shared()`|

------------------------------

```{r}
sessionInfo()
```

